Add cirq.GateFamily and cirq.Gateset

tanujkhattar opened the pull request:
This PR adds two new types - `cirq.GateFamily` and `cirq.Gateset`  to Cirq. It is based on the design proposed in RFC https://tinyurl.com/cirq-gatesets. 

This is the first PR towards the roadmap item #3243 . Follow up PRs will use the newly created `Gateset` type in devices and compilers to replace the existing ad-hoc checks for containment and improve description and validation. 

Note that the `GateFamily` and `Gateset` classes are currently `not_yet_serializable` because of a technical difficulty around serializing cirq types (not instances).  This will be addressed in a later PR.

@MichaelBroughton @dstrain115 @maffoo @Strilanc - PTAL

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        for gate_family in self.gates:
+            if item in gate_family:
+                return True
+        return False
```
Can we make this O(1) instead of O(n) ?

tanujkhattar replied:
Since derived gate families can have custom predicates, it's not possible to make it O(1) for the general case. We could do optimizations specific to the default gate family to check in O(1) for the trivial case and then do an O(n) if no match was found, but I would rather defer such optimizations to the point where we can show that this O(n) is becoming a performance bottleneck. Note that I'd generally expect the number of gate families in a gateset to be fairly small (<= 10 on average?) so the `n` in `O(n)` is more like a small constant. 

Sure, it can lead to a `10x` performance degradation factor when validating a large circuit, but do we already have circuits large enough where this is a bottleneck?

MichaelBroughton replied:
I wouldn't put it past some brave users to create gatesets with as many as 50 or more gates in a gateset. Do you think there might be any way we could keep it O(1) for a vanilla `GateFamily` instance and then special case any of the subclassed `GateFamily` instances ?

tanujkhattar replied:
Made containment O(1) for vanilla `GateFamily` instances. PTAL.

MichaelBroughton reviewed (changes requested):
Thanks for opening the PR, this is a good first pass. I mostly focused on `GateFamily ` since the containment checks for `GateSet` are O(n) I figured we ought to try and cut that down before doing any serious review there. Also, do we think we should put this module in the `ops` directory ? It doesn't really feel like an operation.

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+    def description(self) -> str:
+        return self._description
+
+    def predicate(self, g: raw_types.Gate) -> bool:
+        """Checks whether gate instance `g` belongs to this GateFamily."""
+        return g == self.gate if isinstance(self.gate, raw_types.Gate) else isinstance(g, self.gate)
```
Well, The type of `self.gate` is `Union[Type[raw_types.Gate], raw_types.Gate]`. So if the type of `self.gate` is 

a) `raw_types.Gate`, the method will check for gate comparison (if condition will trigger). 
b) `Type[raw_types.Gate]`, the method will do an `isinstance` check, but it'll pass only if `g` is an instance of `raw_types.Gate` because type of `self.gate` is `Type[raw_types.Gate]`. Hence, the method will not accept arbitrary types. 


The reason I've chosen to support only gate containment check here is to make it harder for users writing new derived gate families to make mistakes. Supporting operations in this method would require that users are aware of the complex operation and gate relationships, which is not always true. I want to avoid users writing checks like `isinstance(op, GateOperation)` for validation, which is currently present all over the library. Ofc, users are free to override the `__contains__` method if they need to handle certain operations in a special way.

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Types defining GateFamily and Gateset"""
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+
+    In order to create gate families with constraints on parameters of a gate
+    type, users should derive from the GateFamily class and override the
+    predicate method used to check for containment.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+    ) -> None:
+        """Init GateFamily
+        Args:
+            gate: A gate type or a non-parameterized instance of a gate type.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+            description: Human readable description of the gate family.
+
+        Raises:
+            ValueError: if `gate` is not a `cirq.Gate` instance or subclass.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+        if isinstance(self.gate, raw_types.Gate):
+            return f'Instance GateFamily: {self._gate_str()}'
+        else:
+            return f'Type GateFamily: {self._gate_str()}'
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+        if isinstance(self.gate, raw_types.Gate):
+            return f'Accepts `cirq.Gate` instances `g` s.t. `g == {self._gate_str()}`.'
+        else:
+            return f'Accepts `cirq.Gate` instances `g` s.t. `isinstance(g, {self._gate_str()})`.'
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+
+    By default, GateFamily supports initialization and containment for
+        a) Non-parameterized `cirq.Gate` instances.
+        b) Gate types deriving from `cirq.Gate`.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,264 @@
+        return self._description
+
+    def predicate(self, g: raw_types.Gate) -> bool:
+        """Checks whether gate instance `g` belongs to this GateFamily."""
```
done

tanujkhattar commented:
I've addressed the comments and left my thoughts around making gateset containment `O(1)`. PTAL. 

Regarding this being in the `ops` directory, since the functionality is so tightly coupled to grouping and validating `gates`, I think it'd be best to keep it under `ops`. The other candidate can be `devices`, but the whole point of an explicit `Gateset` class is to try and decouple the two concepts. This would also be used by `optimizers` and not just `devices`.I'm not sure if we have a better choice than `ops`, but LMK your thoughts.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        self,
+        op: raw_types.Operation,
+        *,
+        unroll_circuit_op: bool = True,
```
When would you want this to be false ?

tanujkhattar replied:
Google backend supports serializing circuit operations and accepting circuits with CircuitOperations, but not all vendors would necessarily support them. Hence, during compilation, turning off the flag would mean that CircuitOperations are not accepted as already present in the target gateset and would continue to decompose into more primitive operations whose serialization is supported.

MichaelBroughton replied:
Makes sense.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        """
+        self._name = name
+        self._gates = frozenset(
+            [g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates]
```
Nit: don't need the square braces here.

tanujkhattar replied:
done

MichaelBroughton reviewed (commented):
I think GateFamily is looking pretty good now. Took a closer look at GateSet, I have some questions around `validate` and `_validate_operation` along with a few minor nits.

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        op: raw_types.Operation,
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
```
The idea is to keep `__contains__` simple enough by delegating the containment checks to the underlying `GateFamily` and having additional helpers like `validate` which try to be smart while handling different types of operations (eg: `GlobalPhaseOperation`, `CircuitOperation` etc.). This is also why `contains` doesn't provide ways to deal with other types of operations like `CircuitOperations`.

I think it's beneficial to keep the smart handling of different operation types decoupled from the basic functionality of checking for containment in the underlying gate families.

MichaelBroughton replied:
That makes sense to me.

Thinking more about this I'm wondering if we can move both: `unroll_circuit_op` and `accept_global_phase` global phase into the constructor. If a user is a given a gateset to work with, they may not immediately know which values these flags should take on in their particular setting and it would be better to fully specify the kinds of support this gateset has during construction.

tanujkhattar replied:
Hmm, that's a good point. How about having the default values of these flags as part of the constructor but also allowing an override of the default values in `validate` ? I think we should allow the flexibility of validating circuits with these options configurable.

MichaelBroughton replied:
I think it would make more sense to just have it all be in the constructor. Giving the user more than one way to specify these options that looks very similar but behave slightly differently just seems like it's giving more ways for things to go wrong in a confusing way.

tanujkhattar replied:
done for now. We can add params to `validate` if we have a strong use case later.

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+            )
+        elif isinstance(op, circuit_operation.CircuitOperation) and unroll_circuit_op:
+            return self.validate(
+                op.mapped_circuit(deep=True),
```
Yes, this will be unroll the circuit `X` times. The reason I used this is because we need to recursively flatten any `CircuitOperation`s present inside the circuit contained by top level `CircuitOperation`. I'm not sure if there's a way to just get a list of all unique operations inside a `CircuitOperation` and avoid any repetitions. 

@95-martin-orion Does such a functionality already exist? If not, Do you think this would be a useful function to add to the CircuitOperation class or should I write a custom unroller here?

95-martin-orion replied:
`op.circuit` contains all unique operations in the `CircuitOperation` modulo measurement key changes. You'll need to manually apply the parameter mapping, though - that can be done with:
```
protocols.resolve_parameters(op.circuit, op.param_resolver, recursive=False)
```
Parameter values are the same for every iteration. We originally considered providing an "iterator" parameter which changed for each repetition, but that feature didn't make it into the released version.

tanujkhattar replied:
I think it would be nice if we can add a method on `CircuitOperation` which allows us to iterate over / returns us a list of all mapped operations without the repetition. 

For eg, even if I apply the `resolve_parameters` mentioned above, the qubit mapping would also need to applied manually at the call site -- and this can affect whether an operation is "valid" or not for topology aware gate families. 

In general, applying CircuitOperation specific mappings outside of the class sounds like a bad idea because if any new kind of mapping is introduced in the future, the call sites would also need to be changes. Thoughts?

tanujkhattar replied:
I've made the change here to explicitly use the underlying `op.circuit` and resolve parameters and apply qubit mappings, but I still think having a utility method on `CircuitOperation` would be the right thing to do here.

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        b) All instances of a specific gate class are allowed.
+
+    Users can define custom gate families for more complex scenarios and
+    use the GateFamily objects as members of a Gateset.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        b) Validating a given gate / optree against the set of allowed gates
+
+    Gatesets rely on the underlying GateFamily's for both description and
+    validation purposes.
```
done

tanujkhattar commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,292 @@
+        description is populated.
+
+        Args:
+            gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
```
done

tanujkhattar commented:
Addressed the comments and made Gateset containment `O(1)` for the vanilla `GateFamily` instances. This is ready for a re-review.

mpharrigan commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,304 @@
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> None:
+        """Init GateFamily
```
I think there's supposed to be a newline between the first summary line and the rest of a docstring

tanujkhattar replied:
Added.

mpharrigan commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,304 @@
+    def description(self) -> str:
+        return self._description
+
+    def predicate(self, gate: raw_types.Gate) -> bool:
```
is there a more obvious name we could use here? `belongs()`?

tanujkhattar replied:
I expect most users to use the `g in gate_family` construct for checking containment. I think the name `predicate` might be more instructive for developers overriding the default predicate by implementing new gate families inheriting from this class. We could also make this method private and have a `def _predicate()`.  

I don't have strong opinions here. LMK your thoughts and I'll be happy to change.

tanujkhattar replied:
Made the predicate a protected member to avoid exposing confusing APIs. Now the users should always check for containment via `g in gate_family`.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,309 @@
+        a) Instance Family:
+            >>> gate_family = cirq.GateFamily(cirq.X)
+            >>> assert (cirq.X in gate_family) == True
+            >>> assert (cirq.X ** sympy.Symbol("theta") not in gate_family) == False
```
Should this be `== True` ?

tanujkhattar replied:
Yeah, thanks for catching this. I think I'll just remove the `==`'s and keep asserts with `in` and `not in` constructs.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,309 @@
+        """
+        g = item if isinstance(item, raw_types.Gate) else item.gate
+        if g in self._instance_gate_families:
+            assert item in self._instance_gate_families[g]
+            return True
+        for gate_mro_type in type(g).mro():
+            if gate_mro_type in self._type_gate_families:
+                assert item in self._type_gate_families[gate_mro_type]
```
Nit: remove assert statements.

tanujkhattar replied:
I think the assert here is important to verify that the internal code is working as expected. It's a tricky implementation and the assert would be useful to safeguard in any future changes we might make.

MichaelBroughton replied:
Fair enough: Can you add a description to them if they get triggered that reads something like: "Match found in GateSet with missing underlying GateFamily entry" for each of the asserts then ?

tanujkhattar replied:
done.

tanujkhattar commented:
@MichaelBroughton Resolved all comments. LMK if this looks good now.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,352 @@
+            *gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
```
Looking through this we aren't really unrolling vs not unrolling circuit operation anymore. We are more like "supporting" vs "not supporting" CircuitOperation. Can we change this to `include/accept_circuit_op` or something similar ?

tanujkhattar replied:
As discussed offline, `include` and `accept` can be more confusing and `unroll` seems to be to indicative of what's happening. Resolving this now.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,352 @@
+            name: (Optional) Name for the Gateset. Useful for description.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
```
Looking closer at this boolean, do we need to special case this here. Wouldn't it make for a simpler API if the user just included a `cirq.GateFamily(cirq.GlobalPhaseOperation)` (type based) when constructing the `GateSet` they want to use instead of having this boolean flag for things ?

tanujkhattar replied:
Well, constructing a `GateFamily` with an operation that doesn't have any underlying "gate" can also be pretty confusing? 

I would prefer not to construct `GateFamilies` that work on any operation type. It opens up a new can of worms and would bring the all the complexities and confusion of Gates vs Operations relationships to the Gateset world as well. 

This is also the reason why `__contains__` in a gateset should be simple. We can fix the the above mentioned confusion using documentation -- use `gateset.validate()` for validating objects. Use containment for checking whether a "gate" (or a "gate" applied to qubits) is contained in the "gateset" or not.

MichaelBroughton replied:
Ahhh god point. Makes a lot of sense to keep here then.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,352 @@
+    def with_params(
+        self,
+        *,
+        name: Optional[str] = None,
+        unroll_circuit_op: Optional[bool] = None,
+        accept_global_phase: Optional[bool] = None,
```
Do we need this new `with_params` as part of the PR ? Can we not just use `**kwargs`  to cut down on redocumenting how to build the object twice ?

tanujkhattar replied:
Yes -- this is used in tests and we can't use **kwargs because we need to use `self.<value>` if the provided `<value>` is None.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,352 @@
+        # To avoid circular import.
+        from cirq.circuits import circuit
+
+        print(
```
Nit: delete.

tanujkhattar replied:
done

MichaelBroughton reviewed (changes requested):
Looking at this more closely, we should probably try and push as much of the functionality from `_validate_operation` into `__contains__` as we can here. Having a user be able to do:

```python3
my_gateset = cirq.GateSet(<stuff>, unroll_circuit_op=True)
my_circuit = cirq.Circuit(cirq.H(...), cirq.CircuitOperation(...))
for op in my_circuit:
  if op not in my_gateset:
    print('Uh oh')

print(my_gateset.validate(my_circuit)) # possible 'True' if constructed with `unroll_circuit_op`
```
and see Uh oh and True output would be very confusing and they might not know which one to trust. If we move most of the logic from `_validate_operation` into `__contains__` the behavior will be more consistent and intuitive across both places. If I could write something like:


```python3
my_gateset = cirq.GateSet(<stuff>, accept_circuit_op=True, accept_tagged_op=False, accept_global_phase_op=False)
my_circuit = cirq.Circuit(cirq.H(...), cirq.CircuitOperation(...))
for op in my_circuit:
  if op not in my_gateset:
    print('Uh oh')

my_gateset.validate(my_circuit) # True
```
And I don't see the "Uh oh" this would be a lot more consistent for the user. This also makes me wonder if we should clarify what of these more "special" kinds of operations (tagged, CircuitOp, GlobalPhase) can be supported by `GateFamily` and which can not be, what happens right now if I put a CircuitOp or GlobalPhase in a `GateFamily` ?

tanujkhattar commented:
@MichaelBroughton Resolved the inconsistency between `validate` and `__contains__` and added tests to verify that they are consistent. Also added lots of documentation to clarify the difference between `validate` and `__contains__`. 

PTAL.

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,378 @@
+            item: The `cirq.Gate` or `cirq.Operation` instance to check containment for.
+        """
+        if isinstance(item, raw_types.Operation) and item.gate is None:
+            return self._validate_operation(item)
```
Should this call into `_validate_operation` or `validate` ?

tanujkhattar replied:
I've assumed here that the user will respect the function signature and not pass in an `op_tree` or `circuit` to check for containment in the Gateset. Hence, calling `_validate_operation` is the right thing to do. 

The recommendation in `_validate_operation` is also that:
"Users should override this method to define custom behavior for operations that do not have an underlying `cirq.Gate`."

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -0,0 +1,378 @@
+
+        Containment checks are handled as follows:
+            a) For Gates or Operations that have an underlying gate (i.e. op.gate is not None):
+                - Forwards the containment check to the underlying GateFamily's
+                - Examples of such operations include `GateOperations` and their controlled and
+                    tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is not None) etc.
+
+            b) For Operations that do not have an underlying gate:
+                - Forwards the containment check to `self._validate_operation(item)`.
+                - Examples of such operations include `CircuitOperations` and their controlled
+                    and tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is None) etc.
```
Missing a few links in here where `TaggedOperation` -> `cirq.TaggedOperation` for docs linking.

tanujkhattar replied:
done

MichaelBroughton reviewed (approved):
LGTM

tanujkhattar commented:
Thanks for the reviews. Merging now.
